examples/waspleau/ext/serverSetup.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * Setup Function
+ * 
+ * This function is what we default export from this module.
+ * It's primary duty is to setup the workers functions, the queue,
+ * and kick off the jobs that will invoke worker functions.
+ * */
+export default async () => {
+  setupWorkerFunctions()
+  await setupQueue()
+  await populateQueue()
+}
+
+
+/**
+ * Dashboard Data
+ * 
+ * This is our in-memory data store we send down to the client via a query in dashboard.js.
+ * This of course means every time the server is restarted we lose this data and must repopulate.
+ * */
+const dashboardData = {}
```

Comment:
Coming from the functional programming world now, I feel uneasy seeing global state :D.

Idea: what if we put `dashboardData` and `workers` and `workerFunctions` in the anonymous function that you are exporting as default export? So they would be created inside it, and then it would return them as its return value? Do you think that would overly complicate things, or it could be an improvement?

Reply:
Ah yes, I agree. It did make me uneasy as well. I can tinker with an approach like that. I also wonder from a JS GC perspective if I can get away with encapsulating by just putting all the variables and functions inside the anonymous function, since `queue.process('*'` has a callback that references `dashboardData` and `workerFunctions` they should not be garbage collected as they will be in the closure? I can test that part out, as well as see if returning them helps/differs any.

Reply:
Ok, so I thought on it a bit more and am stumped on how to get rid of the top level declarations 😄  I think the hard part is the fact I need to access `dashboardData` in the queue callback, but then _also_ expose it to the query via some `export`ed function. I don't know that returning it from the default anonymous function would enable that second part (unless I missed part of the suggestion).

The good news is I think all modules are themselves wrapped in functions, so it is not global to all of node just this file. But if you have ideas for how we can do something like the following in a more locally scoped declaration happy to give it a go:

```javascript
// how can we declare dashboardData such that we use it in the queue callback
// registered in setupFn, but also export a way to get that data for a query?

export default async () => {
  queue.process('*', async (job) => {
    dashboardData[foo] = bar
  })
}

export const getDashboardData = () => Object.values(dashboardData).flat()
```

Thanks!

Reply:
I know for sure that if you put the variable in the anonymous closure, it will live as long as it is being used, GC will not clean it up.

As for modules -> you are right, they are already encapsulated, so these variables are global only inside the module, and not through the whole codebase. Module is a like a big function that gets executed on import. I am not sure though if it can get executed multiple times, in case it is imported in multiple places -> I guess not, that would be problematic. So actually using these variables in this fashion as you did is not so bad, you are right, and probably it is not worth further optimizing and complicating. If I am correct, I believe we can look at JS module as a singleton.

Regarding using dashboardData in query -> you are right! They only way really is to keep some state in the module, and if we are keeping state, then it doesn't really matter if it is one variable or couple of them.

TLDR; You are right, modules are encapsualted / singletons and what you did is actually completely fine and probably the best way!

Reply:
Cool thanks for the review, I appreciate it. I just did not want to miss an approach here, but it was an interesting use case to think about! Thanks